### PR TITLE
Block-SpMM TPP for x86 (avx2, avx512, AMX) and ARM (sve256)

### PR DESCRIPTION
**Pull request by egeor**
This PR introduces Block-SpMM TPP for x86 (avx2, avx512, AMX) and ARM (sve256) (supported precisions: F32, BF16, INT8)

### CODE REVIEW

**Comment by alheinecke** [comment]
Can you please add a unit test?

File: include/libxsmm.h [hunk 1]
```diff
@@ -311,6 +311,9 @@ LIBXSMM_API libxsmm_gemmfunction libxsmm_create_packed_spgemm_csc_v2(
   const libxsmm_gemm_shape gemm_shape, const libxsmm_bitfield gemm_flags, const libxsmm_bitfield prefetch_flags, const libxsmm_blasint packed_width,
   const unsigned int* column_ptr, const unsigned int* row_idx, const void* values);
 
+LIBXSMM_API libxsmm_gemmfunction libxsmm_create_packed_spgemm_bcsc(
+  const libxsmm_gemm_shape gemm_shape, const libxsmm_bitfield gemm_flags, const libxsmm_bitfield prefetch_flags, const libxsmm_blasint packed_width, const libxsmm_blasint bk, const libxsmm_blasint bn);
```
Comment (alheinecke) [inline include/libxsmm.h:315]: Nit: I order to align with the rest of the API, a struct "libxsmm_spgemm_config" would be great for the last 3 args.

File: samples/spmm/spmm_kernel.c [hunk 2]
```diff
@@ -0,0 +1,956 @@
...
+  }
+
+  printf("function pointer address: %llx\n", (unsigned long long)l_test_jit.xmm);
+  printf("%fs for creating jit\n", l_jittime);
+
+  return l_runtime;
+}
+
+LIBXSMM_INLINE
+double get_random_posneg_p5_num(void) {
```
Comment (alheinecke) [inline samples/spmm/spmm_kernel.c:439]: Please create an issue after merging, that we should try to common-nize some of these functions, e.g. they seem to be copied from xgemm.

File: src/generator_aarch64_instructions.c [hunk 3]
```diff
@@ -952,6 +952,7 @@ void libxsmm_aarch64_instruction_sve_move( libxsmm_generated_code*
     case LIBXSMM_AARCH64_INSTR_SVE_LD1H_I_OFF:
     case LIBXSMM_AARCH64_INSTR_SVE_LD1W_SR:
     case LIBXSMM_AARCH64_INSTR_SVE_LD1W_I_OFF:
+    case LIBXSMM_AARCH64_INSTR_SVE_LD1D_V_OFF64:
```
Comment (alheinecke) [inline src/generator_aarch64_instructions.c:955]: please add the instructions to the encoder tester in samples/encoder

File: src/libxsmm_main.c [hunk 4]
```diff
@@ -3578,6 +3635,42 @@ LIBXSMM_API libxsmm_gemmfunction libxsmm_create_packed_spgemm_csc_v2(
   return result.xgemm.gemm;
 }
 
+LIBXSMM_API libxsmm_gemmfunction libxsmm_create_packed_spgemm_bcsc(
+  const libxsmm_gemm_shape gemm_shape, const libxsmm_bitfield gemm_flags, const libxsmm_bitfield prefetch_flags, const libxsmm_blasint packed_width, const libxsmm_blasint bk, const libxsmm_blasint bn)
```
Comment (alheinecke) [inline src/libxsmm_main.c:3639]: see above on the API

File: src/template/libxsmm.h [hunk 5]
```diff
@@ -311,6 +311,9 @@ LIBXSMM_API libxsmm_gemmfunction libxsmm_create_packed_spgemm_csc_v2(
   const libxsmm_gemm_shape gemm_shape, const libxsmm_bitfield gemm_flags, const libxsmm_bitfield prefetch_flags, const libxsmm_blasint packed_width,
   const unsigned int* column_ptr, const unsigned int* row_idx, const void* values);
 
+LIBXSMM_API libxsmm_gemmfunction libxsmm_create_packed_spgemm_bcsc(
+  const libxsmm_gemm_shape gemm_shape, const libxsmm_bitfield gemm_flags, const libxsmm_bitfield prefetch_flags, const libxsmm_blasint packed_width, const libxsmm_blasint bk, const libxsmm_blasint bn);
```
Comment (alheinecke) [inline src/template/libxsmm.h:315]: see above.

**Review by alheinecke — COMMENTED** [review]
Looks very good :-), I have some minor commits:
a) align API, should be very quick
b) add additional aarch64 instructions to encoder
c) file issue for future clean-up after merge

I suggest also removing the sparse_weight_matrix, etc. in samples/deeplearning as they are superseded by parlooper benchmarks, correct?

File: samples/spmm/spmm_kernel.c (hunk 2 above)
Comment (hfp) [reply samples/spmm/spmm_kernel.c:439]: I can take such issue and add an tool able to create certain number of bins on top of RNGs.

Comment (alheinecke) [reply samples/spmm/spmm_kernel.c:439]: let's first review what we need after this is merged

Comment (hfp) [reply samples/spmm/spmm_kernel.c:439]: ACK

**Comment by egeor** [comment]
> I suggest also removing the sparse_weight_matrix, etc. in samples/deeplearning as they are superseded by parlooper benchmarks, correct?

Nope, those ones cover also the plain csc, csr, with hardwired sparsity etc...
The parlooper benchmark covers only the new bcsc kernel.

**Review by alheinecke — APPROVED** [review]

**Comment by alheinecke** [comment]
Fine for now, I'm just wondering in which case the plain csc/csr/hardwired sparsity would be useful == faster than the code we are merging.

**Pull request merged**